### ISSUE TITLE
Folder contains more entries than file_limit (25000)

### ISSUE THREAD

**Issue by josvr** [syncany/syncany#353]
In Use: Syncany 0.4.0-alpha

Hi,

I am using SyncAny to backup some large directories. Now I get the exception: "Folder contains more entries than file_limit (25000)".

I think the following goes wrong:

* Please post instructions to investigate/workarround this issue.
* What is the directory which contains a huge amount of files? This is not logged.Now I cannot see which directory is mentioned. So I cannot login to the Dropbox UI to check what is going on (e.g. files not cleaned up by SyncAny?)
* How much files it does contain is not logged. Maybe this information is not available because the DropBox API simply cancels the operation?
* If this is not a bug in the Dropbox Plugin (e.g. not cleaning up tmp files) then the Dropbox plugin should handle Dropbox limitations vs Filesystem limitations for example by splitting in subdirectories on dropbox side!?

```
23-1-15 8:15:45.368 | InternalEventBu | WR/CloudStorag | INFO : Event bus 'LocalEventBus': Posting event StatusEndSyncExternalEvent (org.syncany.operations.daemon.messages.StatusEndSyncExternalEvent@114af64) ...
23-1-15 8:15:45.370 | WebServer       | WR/CloudStorag | INFO : Sending broadcast message to 0 websocket client(s)
23-1-15 8:15:45.371 | AbstractTransfe | WR/CloudStorag | INFO : Looking for other running remote operations ...
23-1-15 8:15:49.751 | DropboxTransfer | WR/CloudStorag | SEVE : Unable to list Dropbox directory.
com.dropbox.core.DbxException$BadResponseCode: unexpected HTTP status code: 406: {"error": "Folder contains more entries than file_limit (25000)"}
        at com.dropbox.core.DbxRequestUtil.unexpectedStatus(DbxRequestUtil.java:212)
        at com.dropbox.core.DbxClient$2.handle(DbxClient.java:186)
        at com.dropbox.core.DbxRequestUtil.doGet(DbxRequestUtil.java:242)
        at com.dropbox.core.DbxClient.doGet(DbxClient.java:1942)
        at com.dropbox.core.DbxClient.getMetadataWithChildrenBase(DbxClient.java:181)
        at com.dropbox.core.DbxClient.getMetadataWithChildren(DbxClient.java:148)
        at org.syncany.plugins.dropbox.DropboxTransferManager.list(DropboxTransferManager.java:247)
        at org.syncany.plugins.transfer.RetriableTransferManager$8.execute(RetriableTransferManager.java:138)
        at org.syncany.plugins.transfer.RetriableTransferManager.retryMethod(RetriableTransferManager.java:197)
        at org.syncany.plugins.transfer.RetriableTransferManager.list(RetriableTransferManager.java:135)
        at org.syncany.plugins.transfer.TransactionAwareTransferManager.list(TransactionAwareTransferManager.java:138)
        at org.syncany.operations.AbstractTransferOperation.otherRemoteOperationsRunning(AbstractTransferOperation.java:97)
        at org.syncany.operations.up.UpOperation.checkPreconditions(UpOperation.java:341)
        at org.syncany.operations.up.UpOperation.execute(UpOperation.java:124)
        at org.syncany.operations.watch.WatchOperation.runSync(WatchOperation.java:254)
        at org.syncany.operations.watch.WatchOperation.execute(WatchOperation.java:140)
        at org.syncany.operations.daemon.WatchRunner$1.run(WatchRunner.java:83)
        at java.lang.Thread.run(Thread.java:745)
23-1-15 8:15:49.753 | RetriableTransf | WR/CloudStorag | WARN : Transfer method failed. 1/3 retries. Sleeping 3000ms ...
org.syncany.plugins.transfer.StorageException: com.dropbox.core.DbxException$BadResponseCode: unexpected HTTP status code: 406: {"error": "Folder contains more entries than file_limit (25000)"}
        at org.syncany.plugins.dropbox.DropboxTransferManager.list(DropboxTransferManager.java:269)
        at org.syncany.plugins.transfer.RetriableTransferManager$8.execute(RetriableTransferManager.java:138)
        at org.syncany.plugins.transfer.RetriableTransferManager.retryMethod(RetriableTransferManager.java:197)
        at org.syncany.plugins.transfer.RetriableTransferManager.list(RetriableTransferManager.java:135)
        at org.syncany.plugins.transfer.TransactionAwareTransferManager.list(TransactionAwareTransferManager.java:138)
        at org.syncany.operations.AbstractTransferOperation.otherRemoteOperationsRunning(AbstractTransferOperation.java:97)
        at org.syncany.operations.up.UpOperation.checkPreconditions(UpOperation.java:341)
        at org.syncany.operations.up.UpOperation.execute(UpOperation.java:124)
        at org.syncany.operations.watch.WatchOperation.runSync(WatchOperation.java:254)
        at org.syncany.operations.watch.WatchOperation.execute(WatchOperation.java:140)
        at org.syncany.operations.daemon.WatchRunner$1.run(WatchRunner.java:83)
        at java.lang.Thread.run(Thread.java:745)
Caused by: com.dropbox.core.DbxException$BadResponseCode: unexpected HTTP status code: 406: {"error": "Folder contains more entries than file_limit (25000)"}
        at com.dropbox.core.DbxRequestUtil.unexpectedStatus(DbxRequestUtil.java:212)
        at com.dropbox.core.DbxClient$2.handle(DbxClient.java:186)
        at com.dropbox.core.DbxRequestUtil.doGet(DbxRequestUtil.java:242)
        at com.dropbox.core.DbxClient.doGet(DbxClient.java:1942)
        at com.dropbox.core.DbxClient.getMetadataWithChildrenBase(DbxClient.java:181)
        at com.dropbox.core.DbxClient.getMetadataWithChildren(DbxClient.java:148)
        at org.syncany.plugins.dropbox.DropboxTransferManager.list(DropboxTransferManager.java:247)
        ... 11 more
23-1-15 8:15:52.755 | RetriableTransf | WR/CloudStorag | WARN : Retrying method: 1/3 ...
23-1-15 8:15:53.625 | NotificationLis | NotifyThread   | INFO : Waiting for incoming message (10000 ms) ...
23-1-15 8:15:56.479 | DropboxTransfer | WR/CloudStorag | SEVE : Unable to list Dropbox directory.
com.dropbox.core.DbxException$BadResponseCode: unexpected HTTP status code: 406: {"error": "Folder contains more entries than file_limit (25000)"}
        at com.dropbox.core.DbxRequestUtil.unexpectedStatus(DbxRequestUtil.java:212)
        at com.dropbox.core.DbxClient$2.handle(DbxClient.java:186)
        at com.dropbox.core.DbxRequestUtil.doGet(DbxRequestUtil.java:242)
        at com.dropbox.core.DbxClient.doGet(DbxClient.java:1942)
        at com.dropbox.core.DbxClient.getMetadataWithChildrenBase(DbxClient.java:181)
        at com.dropbox.core.DbxClient.getMetadataWithChildren(DbxClient.java:148)
        at org.syncany.plugins.dropbox.DropboxTransferManager.list(DropboxTransferManager.java:247)
        at org.syncany.plugins.transfer.RetriableTransferManager$8.execute(RetriableTransferManager.java:138)
        at org.syncany.plugins.transfer.RetriableTransferManager.retryMethod(RetriableTransferManager.java:197)
        at org.syncany.plugins.transfer.RetriableTransferManager.list(RetriableTransferManager.java:135)
        at org.syncany.plugins.transfer.TransactionAwareTransferManager.list(TransactionAwareTransferManager.java:138)
        at org.syncany.operations.AbstractTransferOperation.otherRemoteOperationsRunning(AbstractTransferOperation.java:97)
        at org.syncany.operations.up.UpOperation.checkPreconditions(UpOperation.java:341)
        at org.syncany.operations.up.UpOperation.execute(UpOperation.java:124)
        at org.syncany.operations.watch.WatchOperation.runSync(WatchOperation.java:254)
        at org.syncany.operations.watch.WatchOperation.execute(WatchOperation.java:140)
        at org.syncany.operations.daemon.WatchRunner$1.run(WatchRunner.java:83)
        at java.lang.Thread.run(Thread.java:745)
2
```

**Comment by pimotte** [comment]
If large is >100GB, then this might an actual problem and not a bug.

Is there any reason you can't go to the dropbox website and look in your syncany folder? I suspect this concerns the multichunks subfolder.

**Comment by josvr** [comment]
Multichuncks files are roughly 4 MB in size. This is not completely true: some files are smaller. My sync folder is roughly 70 GB and according to Dropbox this is also rougly the utilization. 70 GB / 4 MB is somewhere in the 20K files direction. So it looks like this is the issue.

Is there any fix planned in the dropbox plugin for this limitation? With nowadays increasing file sizes (photos, videos), it makes sense to find a solution for this.

Thanks

**Comment by cr0** [comment]
25000 files per folder seems to be a hard limit set by dropbox. Hence we have to use subfolders in order to ensure that it works even for many a few hundred thousand multichunks. Fix in the making...

**Comment by binwiederhier** [comment]
@josvr First of all, thanks for reporting this. I know that i still owe you a response to your e-mail. My apologies for that. I sometimes wait too long to reply and then it's too late. 

@cr0 I'd suggest a structure based on the first 1-2 multichunk ID bytes:

```
muiltichunks/0d/multichunk-0d59aca23caee4202833ec3d3a33ca8f659b0504
muiltichunks/2e/multichunk-2ee20fa5a34efdb61a18daf38752fc6ca3c48830
muiltichunks/44/multichunk-448cee1701721348feaddee461de3133b4f1e7d1
muiltichunks/7e/multichunk-7e89ffed9d8d2d18a9f7802d22413d67728585e7
muiltichunks/80/multichunk-8045d67039fde03fcfc22b59fe580634166e013c
muiltichunks/c8/multichunk-c8334ba822235fcd19fd9e03b8a3d7148fad8984
...
```

That way it's easy to locate them by their ID. Obviously listing becomes a hassle...

**Comment by josvr** [comment]
Ok, thanks let me know when it is fixed (hopefully backwards compatible with 0.4-alpha or a migration tool will be provided, since I am not really happy if I need to drop 70GB from dropbox and upload everything again).

**Comment by binwiederhier** [comment]
Should be easy enough to provide you a small script ttot correct. Linux/Mac?

**Comment by josvr** [comment]
linux thanks

**Comment by binwiederhier** [comment]
@cr0 @pimotte Do you thinkk 1-byte prefix (`multichunks/[PREFIX]/multichunk-...`) is enough? [A quick calculation](https://docs.google.com/spreadsheets/d/1Z1xtaHYkmbT6KSl2Lnvdm02FewUW-GgoptNVvnOs5BI/edit?usp=sharing) that it should hold up a while if we assume an avg. multichunk size of 1 MB. If multichunks are smaller, however, it might not be sufficient.

@josvr Can you run that command on yourr multichunks/ folder? It tells us the average multichunkk size in your folder. On my repos, I always land between 2 and 3 MB: 

```
find ./ -ls | awk '{sum += $7; n++;} END {print sum/n;}'
```

**Comment by josvr** [comment]
I have excluded SyncAny from Dropbox syncing, so I do not have the multichunks on my Linux box.

I would go for 2 or 3 byte prefix.  If you go with 1, you will find sooner or later someone filling a report which hit this limitation again. 

Maybe the prefix can become dynamic (start with 0-9 then 10-99 etc), then you will never hit this issue.

**Comment by binwiederhier** [comment]
The problem with a 2-byte prefix is that there are 32000 possible combinations, so it would have to be hierarchical `multichunks/01/02/multichunk-...`, which makes listing multichunks even more of a hassle. @pimotte Do we do that anywhere (`list(MultichunkRemoteFile.class)`)? I can't remember.

**Comment by pimotte** [comment]
The 1-byte prefix seems a little lacking. (1 byte == two characters, in this case, right?)

The 1-byte prefix seems to be okayish, but since there is a large number of folders, with reasonable certainty, there will be a couple of folders with significantly more chunks than average, so this will break earlier than you expect. 2-byte prefix should probably be enough for at least the next 16 years or so.

**Comment by binwiederhier** [comment]
So 2-byte hierarchical 2-byte prefix then? Like this: `muiltichunks/0d/59/multichunk-0d59aca23caee4202833ec3d3a33ca8f659b0504`

**Comment by cr0** [comment]
Example

```
$ find ~/Public/Dropbox/Apps/syncany-connector/test01/multichunks 
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/19
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/19/9e
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/19/9e/multichunk-199e0ed2c3f2fd6baa9428ec6f622f82e9d639b6
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/28
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/28/99
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/28/99/multichunk-2899f1fabb4ed3eff7377071178899c8589a1d29
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/50
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/50/04
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/50/04/multichunk-5004c2628f0c52b9a59d1409a8813827cdf73cbb
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/60
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/60/e1
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/60/e1/multichunk-60e13243b227a67b7b803f6ff1786d90235b8f35
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/6a
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/6a/1f
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/6a/1f/multichunk-6a1f3e1cfc108c2bb8e1b67ed3947837d0a10dac
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/7b
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/7b/bc
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/7b/bc/multichunk-7bbc0f94fc341bd22cb8d80fd9bee4fb42e9c5bc
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/8b
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/8b/b6
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/8b/b6/multichunk-8bb6e2dd3f351e31aa3a9fcb0f2e28f792ad46a8
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/98
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/98/d2
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/98/d2/multichunk-98d2f21d1e23acacbd45d43f546df09b60369a83
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/9d
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/9d/3a
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/9d/3a/multichunk-9d3a37805fe413c3f151a34102d4b39fd3ed204f
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/9e
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/9e/19
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/9e/19/multichunk-9e1951108a95303bd08b2779e79feff1b5d64bd7
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/bf
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/bf/eb
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/bf/eb/multichunk-bfebf0ed513899d0c2b7f19595bc751817dc6c1a
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/c9
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/c9/e6
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/c9/e6/multichunk-c9e67eb042e719fd60b86d5e6295dec4bd8ede1e
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/db
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/db/48
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/db/48/multichunk-db483d863dff4307e44dfedb7e60dbaf11a1b227
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/db/7e
/Users/chr/Public/Dropbox/Apps/syncany-connector/test01/multichunks/db/7e/multichunk-db7e6462264494f43af58f47a760785f91491e26
```

However, the Dropbox API lists all files in a folder recursively. I don't know if the limit kicks in if we list more than 25000 files even if they are in different subfolders.

**Comment by cr0** [comment]
https://github.com/syncany/syncany-plugin-dropbox/commit/dabc2db89cd3282cbcf00219fb35848169090614 implements a recursive algorithm to list files

**Comment by cr0** [comment]
Since the 25000 files per folder is a limitation of the REST API I don't see a (trivial) solution to move you 70 gig of Dropbox files because every API---either Java or Python---use the said API. You probably don't get around downloading all files using the official APP, move them with a local shell script and then hope that Dropbox is clever enough to realize that only the location has changed.

**Comment by pim** [comment]
(I for sure don’t ;))

**Comment by cr0** [comment]
Ah sorry wrong pim :D 

/cc @pimotte

**Comment by pimotte** [comment]
Yeah, I'm afraid I can't think of a way to easily migrate, since any automated solution would need to list the directory, which the API refuses to do.

**Comment by cr0** [comment]
Ok then.

I'll want to close this issue as soon as possible as we shouldn't waste another user's bandwidth due to running into the same issue. So do we want to folderizer any other files, too?  Is it possible that Syncany will create more than 10000 database or temp files?

**Comment by josvr** [comment]
Is this issue now fixed (because cr0 is talking about closing) and if yes in which version? Migration tooling is not very important: I can simply reinit dropbox and upload 70 Gig again.

**Comment by cr0** [comment]
It *should* be fixed in this branch: https://github.com/syncany/syncany-plugin-dropbox/tree/hotfix/issue353

I hope that I can find some time this evening or at least tomorrow to test if we circumvent running into any API limits. But you can already try it yourself if you're building the plugin on your machine which runs Syncany (see the wiki on how you compile and install plugins).

**Comment by cr0** [comment]
*Seems like we also have to subfolder temp files*

**Comment by josvr** [comment]
Okay, there is also an issue with temp/tmp files. First on my Linux box the /tmp directory get spooled with tmp files, cause a simple "ls" to be take endless. 

Second if I go into Dropbox GUI and then to the root (  apps/syncany-connector/myrootfolder ), I never get the content of this folder anymore. Most likely because there are a lot of never cleaned up temp files in the root folder running into API limtiations. 

Is this a known issue? Why are temp files never/not all been clean up?  Should I raise an another issue for this? 

Can someone please initialize a Dropbox plugin and then "sy up" a lot of data and then see local /tmp and dropbox for tmp files which are not cleaned up?

**Comment by cr0** [comment]
Yeah I suspected that we need to folderize temp items too... I'll implement the changes later..

**Comment by cr0** [comment]
Ok some news:
We're now folderizing temp items and multi chunks, however, since temp items are also database items we cannot use the id field. Therefore, we generate a sha256 hash from the filename and use those bytes to create subfolders.

Remote temp files and empty folders should be automatically cleaned.

@josvr Feel free to test it again.

**Comment by binwiederhier** [comment]
@cr0 Looking good so far. Tiny not-sure-if-justified-remark: There is a method called `StringUtil.toHex()` that we usually use for byte[] to String conversion. :+1:

**Comment by josvr** [comment]
Can someone explain why temp files are not so temporary in syncany. Why can I get +25000 temp files? Is this normal? I expect when  a 'sy up'  finishes (and it copies the database) it also cleans up the temp files. This seems not to be happen!?

**Comment by pimotte** [comment]
I would say that is a bug. If you are talking local temp files, could you make a separate issue for that? If you're talking remote, can you test if it's a dropbox specific issue? If not, again a separate issue would be better to discuss that.

**Comment by cr0** [comment]
@binwiederhier Changed

@josvr Remote tempfiles are removed with the latest commit, at least at the remote side (dropbox)

**Comment by josvr** [comment]
In which release (version number) will this be delivered, and when is it planned to release that release?

**Comment by cr0** [comment]
I do some stress test this evening and then release the dropbox plugin independent of any Syncany release.

**Comment by cr0** [comment]
@binwiederhier wants the try the fix before we merge it. Probably tonight.

**Comment by cr0** [comment]
**on hold** until #366 is implemented.

**Comment by borisxm** [comment]
It seems that the fixes listed in this issue are incompatible with old repositories. I'm have an old repo on dropbox which was created with 0.4.3-alpha. Since then, it was used mostly with 'sy up' command. Recently, I've tried to get a copy of it to another machine, and got an error:
  Command org.syncany.cli.DownCommand@6cc06bf7 FAILED.  java.io.IOException: Decryption/extraction of multichunk ae6709... failed. The multichunk might have been tampered with!

There were no signs, that this file was altered in any way. After some investigation of log files, it was determined, that plugin tries to download this file from repo/multichunks/ae/67/ae6709.... and this file (as well as folder) doesn't exist, but it exists in the multichunks/ae6709... location. To verify my guesses, I've copied that file to the ae/67 directory and 'sy down' command happily accepted it. Now it is clear, that my repo contains both, old and new layout. Tried 'sy cleanup', but it didn't help much.

How one supposed to recover from this situation? I could do this manually, but it will take ages due to a huge number of files.
